### PR TITLE
[Sparse] Add HGNN example

### LINKED ISSUE #4870: [Sparse] HGNN Example
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
Add HGNN example (https://arxiv.org/pdf/1809.09401.pdf).
Closes #4870.

I listed a question and a small suggestion in the code and I'll list here as well:
* Make `DGLGraph.adj()` return a DGL sparse matrix instead of PyTorch sparse matrix, or do we have a convenient conversion function?
* Have an `eye()` function to construct identity sparse matrix.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/sparse/hgnn.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+"""
+Hypergraph Neural Networks (https://arxiv.org/pdf/1809.09401.pdf)
+"""
+import dgl
+import dgl.data
+import dgl.mock_sparse
```
Comment (mufeili) [inline examples/sparse/hgnn.py:6]: Is it necessary to import `dgl.data` and `dgl.mock_sparse`?

File: examples/sparse/hgnn.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+graph = dataset[0]
+# The paper created the hypergraph by "each time one vertex in the
+# graph is selected as the centroid and its connected vertices
+# are used to generate one hyperedge including the centroid
+# itself".  In this case, the incidence matrix of the hypergraph
+# is the same as the adjacency matrix of the original graph (with
+# self-loops).
+# We follow the paper assuming that the rows of the incidence matrix
+# are nodes and the columns are edges.
```
Comment (mufeili) [inline examples/sparse/hgnn.py:23]: An attempt to improve the description from L16-23

```suggestion
The paper created a hypergraph from the original graph. For each node in the original graph, a hyperedge in the hypergraph is created to connect this center node and all neighbors of the center node including itself. In this case, ...
```

File: examples/sparse/hgnn.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+dataset = dgl.data.CoraGraphDataset()
+
+graph = dataset[0]
+# The paper created the hypergraph by "each time one vertex in the
+# graph is selected as the centroid and its connected vertices
+# are used to generate one hyperedge including the centroid
+# itself".  In this case, the incidence matrix of the hypergraph
+# is the same as the adjacency matrix of the original graph (with
+# self-loops).
+# We follow the paper assuming that the rows of the incidence matrix
```
Comment (mufeili) [inline examples/sparse/hgnn.py:22]:
```suggestion
# We follow the paper and assume that the rows of the incidence matrix
```

File: examples/sparse/hgnn.py (hunk 2 above)
Comment (mufeili) [inline examples/sparse/hgnn.py:23]:
```suggestion
# are for nodes and the columns are for edges.
```

File: examples/sparse/hgnn.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+from torchmetrics.functional import accuracy
+import tqdm
+
+dataset = dgl.data.CoraGraphDataset()
+
+graph = dataset[0]
+# The paper created the hypergraph by "each time one vertex in the
+# graph is selected as the centroid and its connected vertices
+# are used to generate one hyperedge including the centroid
+# itself".  In this case, the incidence matrix of the hypergraph
```
Comment (mufeili) [inline examples/sparse/hgnn.py:19]: An attempt to improve the description from L16-19

```suggestion
The paper created a hypergraph from the original graph. For each node in the original graph, a hyperedge in the hypergraph is created to connect this center node and all neighbors of the center node including itself. In this case, ...
```

File: examples/sparse/hgnn.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+# is the same as the adjacency matrix of the original graph (with
+# self-loops).
+# We follow the paper assuming that the rows of the incidence matrix
+# are nodes and the columns are edges.
+
+# QUESTION: can we either (1) easily convert a PyTorch sparse matrix
+# to a DGL sparse matrix, or (2) make g.adj() a DGL sparse matrix
+# in the future?
```
Comment (mufeili) [inline examples/sparse/hgnn.py:27]: L25-27 has been tracked by https://github.com/dmlc/dgl/issues/4876. You may remove these comments.

File: examples/sparse/hgnn.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+adj = dgl.add_self_loop(graph).adj().coalesce()
```
Comment (mufeili) [inline examples/sparse/hgnn.py:28]: Why not use `.edges()` to access `row` and `col`? Also currently `transpose` is `False` for `.adj()`, so `adj.indices()` yields `src, dst` while strictly speaking we need `create_from_coo(dst, src)` for non-symmetric adjacency matrices.

File: examples/sparse/hgnn.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+row, col = adj.indices()
+values = adj.values()
```
Comment (mufeili) [inline examples/sparse/hgnn.py:30]: Since the graph is not weighted, we don't need `values`. By default, 1-valued tensor will be created for non-zero values if not specified.

File: examples/sparse/hgnn.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+    def forward(self, H, X):
+        d_V = H.sum(1)  # node degree
+        d_E = H.sum(0)  # edge degree
+        D_V_invsqrt = dgl.mock_sparse.diag(d_V ** -0.5)  # D_V ** (-1/2)
+        D_E_inv = dgl.mock_sparse.diag(d_E ** -1)  # D_E ** (-1)
+        # Minor suggestion: could we have dgl.mock_sparse.eye(N)?
+        W = dgl.mock_sparse.diag(torch.ones(d_E.shape[0]))
+
+        conv = D_V_invsqrt @ H @ W @ D_E_inv @ H.T @ D_V_invsqrt
```
Comment (mufeili) [inline examples/sparse/hgnn.py:56]: L49-56 can be done in the pre-processing step.

File: examples/sparse/hgnn.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+        X = conv @ self.Theta1(self.dropout(X))
+        X = F.relu(X)
+        X = conv @ self.Theta2(self.dropout(X))
+        return X
+
+
+model = HGNN(X.shape[1], dataset.num_classes)
+opt = torch.optim.Adam(model.parameters(), lr=0.001)
+best_val_acc, best_test_acc = 0, 0
```
Comment (mufeili) [inline examples/sparse/hgnn.py:66]: Maybe not necessary for a demonstration

**Comment by mufeili** [comment]
What test accuracy numbers can you obtain with this script? What are the numbers in the paper? We don't need to match the numbers as long as they look fine.

File: examples/sparse/hgnn.py (hunk 6 above)
Comment (mufeili) [inline examples/sparse/hgnn.py:28]: Maybe we can directly do `A + I` without using `add_self_loop`.

File: examples/sparse/hgnn.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        self.Theta1 = nn.Linear(in_size, hidden_dims)
+        self.Theta2 = nn.Linear(hidden_dims, out_size)
+        self.dropout = nn.Dropout(0.5)
+
+    def forward(self, H, X):
+        d_V = H.sum(1)  # node degree
+        d_E = H.sum(0)  # edge degree
+        D_V_invsqrt = dgl.mock_sparse.diag(d_V ** -0.5)  # D_V ** (-1/2)
+        D_E_inv = dgl.mock_sparse.diag(d_E ** -1)  # D_E ** (-1)
+        # Minor suggestion: could we have dgl.mock_sparse.eye(N)?
```
Comment (mufeili) [inline examples/sparse/hgnn.py:53]: Use `identity`?

File: examples/sparse/hgnn.py (hunk 1 above)
Comment (BarclayII) [reply examples/sparse/hgnn.py:6]: No need to import `dgl.data`, but `dgl.mock_sparse` will throw an error without explicit importing.

File: examples/sparse/hgnn.py (hunk 8 above)
Comment (BarclayII) [reply examples/sparse/hgnn.py:56]: If `H` is an argument in the forward function then technically I cannot put it in the preprocessing step.  So I guess the problem is whether `H` should be in the forward function or in `__init__`.

File: examples/sparse/hgnn.py (hunk 1 above)
Comment (mufeili) [reply examples/sparse/hgnn.py:6]: I see. Perhaps either `from dgl.mock_sparse import xx` or alternatively `import dgl.mock_sparse as xx`. The existing examples follow the former option.

**Comment by BarclayII** [comment]
> What test accuracy numbers can you obtain with this script? What are the numbers in the paper? We don't need to match the numbers as long as they look fine.

The test accuracy is 81.6% which matches the paper.

**Comment by mufeili** [comment]
I'm good in general, though I guess @frozenbugs might have some further suggestions regarding format/style.

File: examples/sparse/hgnn.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+class HGNN(nn.Module):
+    def __init__(self, H, in_size, out_size, hidden_dims=16):
+        super().__init__()
+
+        self.Theta1 = nn.Linear(in_size, hidden_dims)
+        self.Theta2 = nn.Linear(hidden_dims, out_size)
+        self.dropout = nn.Dropout(0.5)
+
+        # Compute the Laplacian with Sparse Matrix API
```
Comment (frozenbugs) [inline examples/sparse/hgnn.py:21]: Follow this style to highlight the sparse API usage?

################################################################################
# (HIGHLIGHT) Take the advantage of DGL sparse APIs to implement the feature
# pre-computation.
################################################################################

File: examples/sparse/hgnn.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    Y_hat = model(X)
+    val_acc = accuracy(Y_hat[val_mask], Y[val_mask])
+    test_acc = accuracy(Y_hat[test_mask], Y[test_mask])
+    return val_acc, test_acc
+
+def load_data():
+    dataset = dgl.data.CoraGraphDataset()
+
+    graph = dataset[0]
+    # The paper created a hypergraph from the original graph. For each node in
```
Comment (frozenbugs) [inline examples/sparse/hgnn.py:55]: Should this be on the top of the paper?
It sounds like a clarification for the whole implementation instead of a piece of code here.

File: examples/sparse/hgnn.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    Y = graph.ndata["label"]
+    train_mask = graph.ndata["train_mask"]
+    val_mask = graph.ndata["val_mask"]
+    test_mask = graph.ndata["test_mask"]
+    return H, X, Y, dataset.num_classes, train_mask, val_mask, test_mask
+
+def main():
+    H, X, Y, num_classes, train_mask, val_mask, test_mask = load_data()
+    model = HGNN(H, X.shape[1], num_classes)
+    opt = torch.optim.Adam(model.parameters(), lr=0.001)
```
Comment (frozenbugs) [inline examples/sparse/hgnn.py:76]: opt -> optimizer

We should either all use opt as abbr. or just use the full word.
I notice many inconsistency in our code and examples.

File: examples/sparse/hgnn.py (hunk 11 above)
Comment (BarclayII) [reply examples/sparse/hgnn.py]: ~~Speaking of cheap neon light banners...~~

**Review by frozenbugs — APPROVED** [review]
